File: MailClientExample/VB/DevExpress.MailClient.Win/Modules/Contacts.vb
Imports System.ComponentModel
Imports System.Text
Imports DevExpress.XtraEditors.Controls
Imports DevExpress.XtraGrid
Imports DevExpress.XtraGrid.Views.Base
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.Utils
Imports DevExpress.MailClient.Win.Forms
Imports DevExpress.XtraGrid.Views.Layout.ViewInfo
Imports System.Collections
Imports DevExpress.XtraGrid.Views.Layout.Events

Namespace DevExpress.MailClient.Win
    Partial Public Class Contacts
        Inherits BaseModule

        Public Overrides ReadOnly Property ModuleName() As String
            Get
                Return My.Resources.ContactsName
            End Get
        End Property
        Public Sub New()
            InitializeComponent()
            EditorHelper.InitPersonComboBox(repositoryItemImageComboBox1)
            gridControl1.DataSource = DataHelper.Contacts
            gridView1.ShowFindPanel()
            InitIndex(DataHelper.Contacts)
        End Sub
        Protected Overrides ReadOnly Property Grid() As DevExpress.XtraGrid.GridControl
            Get
                Return gridControl1
            End Get
        End Property
        Friend Overrides Sub ShowModule(ByVal firstShow As Boolean)
            MyBase.ShowModule(firstShow)
            gridControl1.Focus()
            UpdateActionButtons()
            If firstShow Then
                gridControl1.ForceInitialize()
                GridHelper.SetFindControlImages(gridControl1)
                If DataHelper.Contacts.Count = 0 Then
                    UpdateCurrentContact()
                End If
            End If
        End Sub
        Private Sub UpdateActionButtons()
            OwnerForm.EnableLayoutButtons(gridControl1.MainView IsNot layoutView1)
            OwnerForm.EnableZoomControl(gridControl1.MainView IsNot layoutView1)
        End Sub
        Private ReadOnly Property CurrentContact() As Contact
            Get
                Return TryCast(CType(gridControl1.MainView, ColumnView).GetFocusedRow(), Contact)
            End Get
        End Property
        Private Sub gridView1_ColumnFilterChanged(ByVal sender As Object, ByVal e As EventArgs) Handles gridView1.ColumnFilterChanged
            UpdateCurrentContact()
        End Sub
        Private Sub gridView1_FocusedRowObjectChanged(ByVal sender As Object, ByVal e As FocusedRowObjectChangedEventArgs) Handles gridView1.FocusedRowObjectChanged
            If e.FocusedRowHandle = GridControl.AutoFilterRowHandle Then
                gridView1.FocusedColumn = colName
            ElseIf e.FocusedRowHandle >= 0 Then
                gridView1.FocusedColumn = Nothing
            End If
            UpdateCurrentContact()
        End Sub
        Private Sub UpdateCurrentContact()
            ucContactInfo1.Init(CurrentContact, Nothing)
            If gridControl1.MainView Is gridView1 Then
                gridView1.MakeRowVisible(gridView1.FocusedRowHandle)
            End If
            OwnerForm.EnableEditContact(CurrentContact IsNot Nothing)
        End Sub
        Protected Friend Overrides Sub ButtonClick(ByVal tag As String)
            Select Case tag
                Case TagResources.ContactList
                    UpdateMainView(gridView1)
                    ClearSortingAndGrouping()
                Case TagResources.ContactAlphabetical
                    UpdateMainView(gridView1)
                    ClearSortingAndGrouping()
                    colName.Group()
                Case TagResources.ContactByState
                    UpdateMainView(gridView1)
                    ClearSortingAndGrouping()
                    colState.Group()
                    colCity.SortOrder = DevExpress.Data.ColumnSortOrder.Ascending
                Case TagResources.ContactCard
                    UpdateMainView(layoutView1)
                Case TagResources.FlipLayout
                    layoutControl1.Root.FlipLayout()
                Case TagResources.ContactDelete
                    If CurrentContact Is Nothing Then
                        Return
                    End If
                    Dim index As Integer = gridView1.FocusedRowHandle
                    gridControl1.MainView.BeginDataUpdate()
                    Try
                        DataHelper.Contacts.Remove(CurrentContact)
                    Finally
                        gridControl1.MainView.EndDataUpdate()
                    End Try
                    If index > gridView1.DataRowCount - 1 Then
                        index -= 1
                    End If
                    gridView1.FocusedRowHandle = index
                    ShowInfo(gridView1)
                Case TagResources.ContactNew
                    Dim contact As New Contact()
                    If EditContact(contact) = DialogResult.OK Then
                        gridControl1.MainView.BeginDataUpdate()
                        Try
                            DataHelper.Contacts.Add(contact)
                        Finally
                            gridControl1.MainView.EndDataUpdate()
                        End Try
                        Dim view As ColumnView = TryCast(gridControl1.MainView, ColumnView)
                        If view IsNot Nothing Then
                            GridHelper.GridViewFocusObject(view, contact)
                            ShowInfo(view)
                        End If
                    End If
                Case TagResources.ContactEdit
                    EditContact(CurrentContact)
            End Select
            UpdateCurrentContact()
            UpdateInfo()
        End Sub
        Private Sub UpdateMainView(ByVal view As ColumnView)
            Dim isGrid As Boolean = TypeOf view Is GridView
            gridControl1.MainView = view
            splitterItem1.Visibility = If(isGrid, DevExpress.XtraLayout.Utils.LayoutVisibility.Always, DevExpress.XtraLayout.Utils.LayoutVisibility.Never)
            layoutControlItem2.Visibility = If(isGrid, DevExpress.XtraLayout.Utils.LayoutVisibility.Always, DevExpress.XtraLayout.Utils.LayoutVisibility.Never)
            GridHelper.SetFindControlImages(gridControl1)
            UpdateActionButtons()
        End Sub
        Private Sub ClearSortingAndGrouping()
            gridView1.ClearGrouping()
            gridView1.ClearSorting()
        End Sub
        Protected Overrides ReadOnly Property AllowZoomControl() As Boolean
            Get
                Return True
            End Get
        End Property
        Public Overrides Property ZoomFactor() As Single
            Get
                Return ucContactInfo1.ZoomFactor
            End Get
            Set(ByVal value As Single)
                ucContactInfo1.ZoomFactor = value
                SetZoomCaption()
            End Set
        End Property
        Protected Overrides Sub SetZoomCaption()
            MyBase.SetZoomCaption()
            If ZoomFactor = 1 Then
                OwnerForm.ZoomManager.SetZoomCaption(My.Resources.Picture100Zoom)
            End If
        End Sub

        Private Sub gridView1_RowCellClick(ByVal sender As Object, ByVal e As RowCellClickEventArgs) Handles gridView1.RowCellClick
            If e.Button = MouseButtons.Left AndAlso e.RowHandle >= 0 AndAlso e.Clicks = 2 Then
                EditContact(CurrentContact)
            End If
        End Sub
        Private Sub gridView1_KeyDown(ByVal sender As Object, ByVal e As KeyEventArgs) Handles gridView1.KeyDown
            If e.KeyData = Keys.Enter AndAlso gridView1.FocusedRowHandle >= 0 Then
                EditContact(CurrentContact)
            End If
        End Sub
        Private Sub layoutView1_CardClick(ByVal sender As Object, ByVal e As CardClickEventArgs) Handles layoutView1.CardClick
            If e.Button = MouseButtons.Left AndAlso e.RowHandle >= 0 AndAlso e.Clicks = 2 Then
                EditContact(CurrentContact)
                layoutView1.UpdateCurrentRow()
            End If
        End Sub
        Private Function EditContact(ByVal contact As Contact) As DialogResult
            If contact Is Nothing Then
                Return DialogResult.Ignore
            End If
            Dim ret As DialogResult = DialogResult.Cancel
            Cursor.Current = Cursors.WaitCursor
            Using frm As New frmEditContact(contact, OwnerForm.Ribbon)
                ret = frm.ShowDialog(OwnerForm)
            End Using
            UpdateCurrentContact()
            Cursor.Current = Cursors.Default
            Return ret
        End Function
        Private Sub UpdateInfo()
            ShowInfo(TryCast(gridControl1.MainView, ColumnView))
        End Sub
        Private Sub layoutView1_ColumnFilterChanged(ByVal sender As Object, ByVal e As EventArgs) Handles layoutView1.ColumnFilterChanged
            UpdateInfo()
        End Sub
        Private Sub repositoryItemHyperLinkEdit1_OpenLink(ByVal sender As Object, ByVal e As OpenLinkEventArgs) Handles repositoryItemHyperLinkEdit1.OpenLink
            If e.EditValue IsNot Nothing Then
                e.EditValue = "mailto:" & e.EditValue.ToString()
            End If
        End Sub
        Protected Sub InitIndex(ByVal list As List(Of Contact))
            SetupGrid(Generate(list, Function(c) AlphaIndex.Group(c.LastName)), indexGridControl)
        End Sub
        Public Sub SetupGrid(ByVal list As List(Of AlphaIndex), ByVal grid As GridControl)
            Dim view As GridView = TryCast(grid.MainView, GridView)
            view.Columns.AddVisible("Index")
            grid.DataSource = list
            AddHandler view.FocusedRowChanged, AddressOf view_FocusedRowChanged
        End Sub
        Private alphaChange As Timer = Nothing
        Private Sub view_FocusedRowChanged(ByVal sender As Object, ByVal e As FocusedRowChangedEventArgs)
            If alphaChange IsNot Nothing Then
                alphaChange.Dispose()
            End If
            alphaChange = New Timer()
            alphaChange.Interval = 200
            AddHandler alphaChange.Tick, Sub(s, ea) alphaChangeTick(s, ea, sender)
            alphaChange.Start()
        End Sub
        Private Sub alphaChangeTick(ByVal s As Object, ByVal ea As Object, ByVal sender As Object)
            CType(s, Timer).Stop()
            Dim focusedIndex = TryCast((CType(sender, GridView)).GetFocusedRow(), AlphaIndex)
            gridControl1.DataSource = ApplyFilter(DataHelper.Contacts, focusedIndex)
            CType(s, Timer).Dispose()
            Me.alphaChange = Nothing
            UpdateInfo()
        End Sub
        Private Function ApplyFilter(ByVal list As List(Of Contact), ByVal alpha As AlphaIndex) As IList
            If alpha Is AlphaIndex.All OrElse alpha Is Nothing Then
                Return list
            End If
            Return list.Where(Function(c) alpha.Match(AlphaIndex.Group(c.LastName))).ToList()
        End Function
        Public Function Generate(ByVal values As List(Of Contact), ByVal extractName As Func(Of Contact, String)) As List(Of AlphaIndex)
            Dim data = _
                From q In values _
                Where extractName(q) IsNot Nothing _
                Group q By GroupKey = extractName(q) Into g = Group _
                Order By GroupKey _
                Select New AlphaIndex() With {.Index = GroupKey, .Count = g.Count()}
            Dim res = data.ToList()
            res.Insert(0, AlphaIndex.All)
            Return res
        End Function
    End Class

    Public Class AlphaIndex
        Public Property Index() As String
        Public Property Count() As Integer
        Public Overrides Function ToString() As String
            Return String.Format("{0}: {1}", Index, Count)
        End Function
        Public Function Match(ByVal text As String) As Boolean
            Return Group(text) = Index
        End Function
        Public Shared Function Group(ByVal text As String) As String
            If String.IsNullOrEmpty(text) Then
                Return Nothing
            End If
            If text.Length > 1 Then
                text = text.Substring(0, 1)
            End If
            If text.Length = 1 Then
                If Char.IsNumber(text.Chars(0)) Then
                    Return "0-9"
                End If
            End If
            Return text.ToUpper()
        End Function
        Private Shared all_ As AlphaIndex
        Public Shared ReadOnly Property All() As AlphaIndex
            Get
                If all_ Is Nothing Then
                    all_ = New AlphaIndex() With {.Count = 0, .Index = "ALL"}
                End If
                Return all_
            End Get
        End Property
        Private Shared alphaNumber_ As AlphaIndex
        Public Shared ReadOnly Property AlphaNumber() As AlphaIndex
            Get
                If alphaNumber_ Is Nothing Then
                    alphaNumber_ = New AlphaIndex() With {.Count = 0, .Index = "0-9"}
                End If
                Return alphaNumber_
            End Get
        End Property
    End Class
End Namespace

File: MailClientExample/CS/DevExpress.MailClient.Win/Modules/Contacts.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Windows.Forms;
using DevExpress.MailClient.Win.Forms;
using DevExpress.XtraEditors.Controls;
using DevExpress.XtraGrid;
using DevExpress.XtraGrid.Views.Base;
using DevExpress.XtraGrid.Views.Grid;

namespace DevExpress.MailClient.Win {
    public partial class Contacts : BaseModule {
        public override string ModuleName { 
            get { return Properties.Resources.ContactsName; } 
        }
        //
        public Contacts() {
            InitializeComponent();
            EditorHelper.InitPersonComboBox(repositoryItemImageComboBox1);
            gridControl1.DataSource = DataHelper.Contacts;
            gridView1.ShowFindPanel();
            InitIndex(DataHelper.Contacts);
        }
        protected override DevExpress.XtraGrid.GridControl Grid { 
            get { return gridControl1; } 
        }
        protected override bool AllowZoomControl {
            get { return true; }
        }
        public override float ZoomFactor {
            get { return ucContactInfo1.ZoomFactor; }
            set {
                ucContactInfo1.ZoomFactor = value;
                SetZoomCaption();
            }
        }
        protected override void SetZoomCaption() {
            base.SetZoomCaption();
            if(ZoomFactor == 1)
                OwnerForm.ZoomManager.SetZoomCaption(Properties.Resources.Picture100Zoom);
        }
        internal override void ShowModule(bool firstShow) {
            base.ShowModule(firstShow);
            gridControl1.Focus();
            UpdateActionButtons();
            if(firstShow) {
                gridControl1.ForceInitialize();
                GridHelper.SetFindControlImages(gridControl1);
                if(DataHelper.Contacts.Count == 0) UpdateCurrentContact();
            }
        }
        void UpdateActionButtons() {
            OwnerForm.EnableLayoutButtons(gridControl1.MainView != layoutView1);
            OwnerForm.EnableZoomControl(gridControl1.MainView != layoutView1);
        }
        Contact CurrentContact {
            get { return ((ColumnView)gridControl1.MainView).GetFocusedRow() as Contact; }
        }
        void gridView1_ColumnFilterChanged(object sender, EventArgs e) {
            UpdateCurrentContact();
        }
        void gridView1_FocusedRowObjectChanged(object sender, FocusedRowObjectChangedEventArgs e) {
            if(e.FocusedRowHandle == GridControl.AutoFilterRowHandle)
                gridView1.FocusedColumn = colName;
            else if(e.FocusedRowHandle >= 0)
                gridView1.FocusedColumn = null;
            UpdateCurrentContact();
        }
        void UpdateCurrentContact() {
            ucContactInfo1.Init(CurrentContact, null);
            if(gridControl1.MainView == gridView1)
                gridView1.MakeRowVisible(gridView1.FocusedRowHandle);
            OwnerForm.EnableEditContact(CurrentContact != null);
        }
        protected internal override void ButtonClick(string tag) {
            switch(tag) {
                case TagResources.ContactList:
                    UpdateMainView(gridView1);
                    ClearSortingAndGrouping();
                    break;
                case TagResources.ContactAlphabetical:
                    UpdateMainView(gridView1);
                    ClearSortingAndGrouping();
                    colName.Group();
                    break;
                case TagResources.ContactByState:
                    UpdateMainView(gridView1);
                    ClearSortingAndGrouping();
                    colState.Group();
                    colCity.SortOrder = DevExpress.Data.ColumnSortOrder.Ascending;
                    break;
                case TagResources.ContactCard:
                    UpdateMainView(layoutView1);
                    break;
                case TagResources.FlipLayout:
                    layoutControl1.Root.FlipLayout();
                    break;
                case TagResources.ContactDelete:
                    if(CurrentContact == null) return;
                    int index = gridView1.FocusedRowHandle;
                    gridControl1.MainView.BeginDataUpdate();
                    try {
                        DataHelper.Contacts.Remove(CurrentContact);
                    }
                    finally {
                        gridControl1.MainView.EndDataUpdate();
                    }
                    if(index > gridView1.DataRowCount - 1) index--;
                    gridView1.FocusedRowHandle = index;
                    ShowInfo(gridView1);
                    break;
                case TagResources.ContactNew:
                    Contact contact = new Contact();
                    if(EditContact(contact) == DialogResult.OK) {
                        gridControl1.MainView.BeginDataUpdate();
                        try {
                            DataHelper.Contacts.Add(contact);
                        }
                        finally {
                            gridControl1.MainView.EndDataUpdate();
                        }
                        ColumnView view = gridControl1.MainView as ColumnView;
                        if(view != null) {
                            GridHelper.GridViewFocusObject(view, contact);
                            ShowInfo(view);
                        }
                    }
                    break;
                case TagResources.ContactEdit:
                    EditContact(CurrentContact);
                    break;
            }
            UpdateCurrentContact();
            UpdateInfo();
        }
        void UpdateMainView(ColumnView view) {
            bool isGrid = view is GridView;
            gridControl1.MainView = view;
            splitterItem1.Visibility = isGrid ? DevExpress.XtraLayout.Utils.LayoutVisibility.Always : DevExpress.XtraLayout.Utils.LayoutVisibility.Never;
            layoutControlItem2.Visibility = isGrid ? DevExpress.XtraLayout.Utils.LayoutVisibility.Always : DevExpress.XtraLayout.Utils.LayoutVisibility.Never;
            GridHelper.SetFindControlImages(gridControl1);
            UpdateActionButtons();
        }
        void ClearSortingAndGrouping() {
            gridView1.ClearGrouping();
            gridView1.ClearSorting();
        }
        void gridView1_RowCellClick(object sender, RowCellClickEventArgs e) {
            if(e.Button == MouseButtons.Left && e.RowHandle >= 0 && e.Clicks == 2)
                EditContact(CurrentContact);
        }
        void gridView1_KeyDown(object sender, KeyEventArgs e) {
            if(e.KeyData == Keys.Enter && gridView1.FocusedRowHandle >= 0)
                EditContact(CurrentContact);
        }
        void layoutView1_CardClick(object sender, XtraGrid.Views.Layout.Events.CardClickEventArgs e) {
            if(e.Button == MouseButtons.Left && e.RowHandle >= 0 && e.Clicks == 2) {
                EditContact(CurrentContact);
                layoutView1.UpdateCurrentRow();
            }
        }
        DialogResult EditContact(Contact contact) {
            if(contact == null)
                return DialogResult.Ignore;
            DialogResult result = DialogResult.Cancel;
            Cursor.Current = Cursors.WaitCursor;
            using(frmEditContact frm = new frmEditContact(contact, OwnerForm.Ribbon))
                result = frm.ShowDialog(OwnerForm);
            UpdateCurrentContact();
            Cursor.Current = Cursors.Default;
            return result;
        }
        void UpdateInfo() {
            ShowInfo(gridControl1.MainView as ColumnView);
        }
        void layoutView1_ColumnFilterChanged(object sender, EventArgs e) {
            UpdateInfo();
        }
        void repositoryItemHyperLinkEdit1_OpenLink(object sender, OpenLinkEventArgs e) {
            if(e.EditValue != null) e.EditValue = "mailto:" + e.EditValue.ToString();
        }
        protected void InitIndex(List<Contact> list) {
            SetupGrid(Generate(list, c => AlphaIndex.Group(c.LastName)), indexGridControl);
        }
        public void SetupGrid(List<AlphaIndex> list, GridControl grid) {
            GridView view = grid.MainView as GridView;
            view.Columns.AddVisible("Index");
            grid.DataSource = list;
            view.FocusedRowChanged += view_FocusedRowChanged;
        }
        Timer alphaChange = null;
        void view_FocusedRowChanged(object sender, FocusedRowChangedEventArgs e) {
            if(alphaChange != null)
                alphaChange.Dispose();
            alphaChange = new Timer();
            alphaChange.Interval = 200;
            alphaChange.Tick += (s, ea) =>
            {
                ((Timer)s).Stop();
                var focusedIndex = ((GridView)sender).GetFocusedRow() as AlphaIndex;
                gridControl1.DataSource = ApplyFilter(DataHelper.Contacts, focusedIndex);
                ((Timer)s).Dispose();
                this.alphaChange = null;
                UpdateInfo();
            };
            alphaChange.Start();
        }
        IList ApplyFilter(List<Contact> list, AlphaIndex alpha) {
            if(alpha == AlphaIndex.All || alpha == null)
                return list;
            return list.Where(c => alpha.Match(AlphaIndex.Group(c.LastName))).ToList();
        }
        public List<AlphaIndex> Generate(List<Contact> values, Func<Contact, string> extractName) {
            var data = from q in values
                       where extractName(q) != null
                       group q by extractName(q) into g
                       orderby g.Key
                       select new AlphaIndex() { Index = g.Key, Count = g.Count() };
            var res = data.ToList();
            res.Insert(0, AlphaIndex.All);
            return res;
        }
    }
    public class AlphaIndex {
        public string Index { get; set; }
        public int Count { get; set; }
        public override string ToString() {
            return string.Format("{0}: {1}", Index, Count);
        }
        public bool Match(string text) {
            return Group(text) == Index;
        }
        public static string Group(string text) {
            if(string.IsNullOrEmpty(text))
                return null;
            if(text.Length > 1)
                text = text.Substring(0, 1);
            if(text.Length == 1) 
                if(Char.IsNumber(text[0])) return "0-9";
            return text.ToUpper();
        }
        static AlphaIndex all;
        public static AlphaIndex All {
            get {
                if(all == null)
                    all = new AlphaIndex() { Count = 0, Index = "ALL" };
                return all;
            }
        }
        static AlphaIndex alphaNumber;
        public static AlphaIndex AlphaNumber {
            get {
                if(alphaNumber == null)
                    alphaNumber = new AlphaIndex() { Count = 0, Index = "0-9" };
                return alphaNumber;
            }
        }
    }
}
